fpm start-project command added

Heulitig opened the pull request:
- Added a subcommand "start-project" to create a new template ftd project with a given project name and path
- Subcommand format - **fpm start-project [project-name] [project-path]**
- Options allowed:
  -n or --name = Project Name 
  -p or --path = Project path (relative with the current directory from which this command is invoked)
- **[project-name] is required field** with **Optional [project-path]** (Default path is the current directory '.' )
- Sample usage 
fpm start-project -n sample_project
fpm start-project --name tutorial_app --path projects/tutorials
- Link to the command documentation in fpm.dev - https://github.com/FifthTry/fpm.dev/pull/34

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,56 @@
+
+	fp.write_all(content.as_bytes())
+		.expect("Could not write contents into the file!");
+}
```
This function declares itself as `async` but is not actually async. Use `tokio::fs` for doing async file system operations.

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,56 @@
+use std::fs;
+use std::env;
+use std::fs::File;
+use std::io::prelude::*;
+
```
We do not use `use` statement in FifthTry, look at other files. Always use full name of anything. I was looking at line number 15, say `File::create` and I did not know if it is `std::fs::File` or `tokio::fs::File`, I had to jump to the top, and this time I got luck as this is a new file and the use statement was in the PR diff. In other PRs, I would have to click on view file to see what you meant when you said `File`. 

Code is written once, and read thousands of times. Do not optimise for writer, optimise for reader.

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,56 @@
+    final_dir = base_path/project_folder => template-files
+	*/
+	let base_path = env::current_dir()?.to_str().unwrap().to_string();
+	/* Not using config for base path as it throws error and requires manifest or FPM.ftd file
+	   and since this command should work from anywhere within the system
+	   so we dont need any other file dependency for using it
+	*/
```
Don't use block comment, use line comment only.

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,56 @@
+	*/
+	// let base_path = config.root.as_str();
+
+	let mut final_dir = "".to_string();
+	if let Some(relative_path) = path{
+		match relative_path {
+			"." => final_dir = format!("{}/{}", base_path, name.unwrap()),
+			_ => final_dir = format!("{}/{}/{}", base_path, relative_path, name.unwrap())
+		}
+	}
```
```suggestion
	let final_dir = if let Some(relative_path) = path{
		match relative_path {
			"." => format!("{}/{}", base_path, name.unwrap()),
			_ => format!("{}/{}/{}", base_path, relative_path, name.unwrap())
		}
	};
```

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,56 @@
+		}
+	}
+	// Create all directories if not present
+	fs::create_dir_all(final_dir.as_str())?;
```
Convert everything to `tokio::fs`.

amitu commented:
@Heulitig do not create a fork. You have push access to original repo, in the original repo create and send branches.

amitu commented:
If you are looking for rationale of why: when I do a git pull I want all branches on my machine. If you maintain your fork I have to extra work to get your branch on my machine.

amitu commented:
@Heulitig any comment that you have taken care of, please click on the resolve button also. Also when all comments are resolved either assign it again to me, or post a comment saying "all done, please review again" or something, so I know when to review again.

Heulitig commented:
Done with the changes. Please review it again
- Used tokio::fs instead of std::fs
- Removed unnecessary comments and refactored the required ones (avoid use of multi-line comments)
- Removed the usage of 'use'

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,45 @@
+
+    // name is a required field so it will always be some defined string (cant be None)
+    // path is an optional field and has a default value "." for current working directory (also cant be None)
+    let (name, path) = (name.unwrap(), path.unwrap());
```
This is not right, you should only use unwrap() if it something you are reasonably sure is impossible. If you need some data, do not accept `Option<>`, change the signature to `start_project(name: &str, path: &str)`.

Acception Option, only if it is valid. Here by doing an unwrap() you are saying, "hey, my program will crash if this parameter is not passed". In which case why are you saying something is optional.

amitu replied:
Seems you are calling start project with optional stuff because that is what that argument returns. In which case keep the function signature as Option, but instead of `.unwrap()`, do a match, and if any requirement argument is missing, print an error message, and return from the function. Do not crash. `.unwrap()` means crash, and crash means bug. If user did a mistake, print a user friendly error message.

amitu replied:
Seems you are calling start project with optional stuff because that is what that argument returns. In which case keep the function signature as Option, but instead of `.unwrap()`, do a match, and if any requirement argument is missing, print an error message, and return from the function. Do not crash. `.unwrap()` means crash, and crash means bug. If user did a mistake, print a user friendly error message.

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,45 @@
+
+    // name is a required field so it will always be some defined string (cant be None)
+    // path is an optional field and has a default value "." for current working directory (also cant be None)
+    let (name, path) = (name.unwrap(), path.unwrap());
```
This is not right, you should only use unwrap() if it something you are reasonably sure is impossible. If you need some data, do not accept `Option<>`, change the signature to `start_project(name: &str, path: &str)`.

Acception Option, only if it is valid. Here by doing an unwrap() you are saying, "hey, my program will crash if this parameter is not passed". In which case why are you saying something is optional.

Heulitig replied:
Okay so I can handle each possible cases of data i.e (Some(x) and None) while parsing for both name and path arguments seperately instead of unwraping (which might cause panic if None value are found during runtime) before passing those as functional arguments as standard &str rather than using Option<&str>. Is this accurate or am I wrong somewhere ?

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,45 @@
+    // path is an optional field and has a default value "." for current working directory (also cant be None)
+    let (name, path) = (name.unwrap(), path.unwrap());
+    let final_dir = match path {
+        "." => format!("{}/{}", base_path, name),
+        _ => format!("{}/{}/{}", base_path, path, name),
```
Nopes, this will not work on windows. Always construct file names using `.join()` etc. Checkout `camino` crate.

Heulitig replied:
Okay I will use this

amitu commented on src/commands/start_project.rs:
```diff
@@ -0,0 +1,45 @@
+    // path is an optional field and has a default value "." for current working directory (also cant be None)
+    let (name, path) = (name.unwrap(), path.unwrap());
+    let final_dir = match path {
+        "." => format!("{}/{}", base_path, name),
+        _ => format!("{}/{}/{}", base_path, path, name),
```
Nopes, this will not work on windows. Always construct file names using `.join()` etc. Checkout `camino` crate.

amitu commented on src/main.rs:
```diff
@@ -90,6 +101,33 @@ fn app(authors: &'static str, version: &'static str) -> clap::App<'static, 'stat
+                        .short("p")
+                        .long("path")
+                        .takes_value(true)
+                        .default_value(".")
```
The default is wrong. If the package name is `foo.com/bar`, the default folder name will be `foo.com/bar`.

Heulitig replied:
Okay package names can be urls, got it. Am I allowed to handle this default case manually instead of using the default_value() which is available in clap ?

amitu replied:
I am not sure if you can implement this requirement using `default_value()` alone. Do it in your handler function.

amitu commented on src/main.rs:
```diff
@@ -90,6 +101,33 @@ fn app(authors: &'static str, version: &'static str) -> clap::App<'static, 'stat
+                        .short("p")
+                        .long("path")
+                        .takes_value(true)
+                        .default_value(".")
```
The default is wrong. If the package name is `foo.com/bar`, the default folder name will be `foo.com/bar`.

amitu commented:
@Heulitig create a PR in fpm.dev which documents this command also, and link to that PR in the description of this PR, and this PR into that one.

amitu commented:
@Heulitig create a PR in fpm.dev which documents this command also, and link to that PR in the description of this PR, and this PR into that one.

Heulitig commented:
> @Heulitig create a PR in fpm.dev which documents this command also, and link to that PR in the description of this PR, and this PR into that one.

Add reference links for both PR's

Arpita-Jaiswal commented on src/commands/mod.rs:
```diff
@@ -8,3 +8,4 @@ pub(crate) mod stop_tracking;
 pub(crate) mod sync;
 pub(crate) mod translation_status;
 pub(crate) mod update;
+pub(crate) mod start_project;
```
Always add new-line in the end of the file

Arpita-Jaiswal commented on src/main.rs:
```diff
@@ -2,6 +2,19 @@
+    // Block of code to run when start-project subcommand is used
+    if let Some(project) = matches.subcommand_matches("start-project") {
+        // project-name => required field (any package Url or standard project name)
+        if let Some(name) = project.value_of("package-name") {
+            // project path => optional
+            let path = project.value_of("package-path");
+            fpm::start_project(name, path).await?;
```
Call  `unwrap()` for `package-name` as this is a required field taken care of by the `clap`

Arpita-Jaiswal commented:
The [documentation pull request](https://github.com/FifthTry/fpm.dev/pull/34) has been attached sucessfully.

Arpita-Jaiswal commented:
Approved! Code is up to date with the [documentation](https://github.com/FifthTry/fpm.dev/pull/34)
